### PR TITLE
http: forwarding x-forwarded-proto from trusted proxies

### LINKED ISSUE #4496: x-forwarded-proto gets overwritten when use_remote_address=true
**Bug Template**

*Title*: *x-forwarded-proto gets overwritten when use_remote_address=true*

*Description*:
When Envoy is deployed behind a trusted edge load balancer (in my case, it's provided by cloud vendor Aliyun.com), and correspondingly set `xff_num_trusted_hops` greater than 0, it should forward header `x-forwarded-proto` from downstream to upstream without modification. Even with `use_remote_address` set to true.

*Repro steps*:

See: https://github.com/5long/envoy-forwarded-proto with a working Docker Stack config to reproduce it.

*Config*:
It's included in the git repo above.

### PR DESCRIPTION
**Pull request by alyssawilk**
Trusting the x-forwarded-proto header from trusted proxies.
If Envoy is operating as an edge proxy but has a trusted hop in front, the trusted proxy should be allowed to set x-forwarded-proto and its x-forwarded-proto should be preserved.
Guarded by envoy.reloadable_features.trusted_forwarded_proto, default on.

Risk Level: Medium (L7 header changes) but guarded
Testing: new unit tests
Docs Changes: n/a
Release Notes: inline
Fixes #4496

### CODE REVIEW

**Comment by alyssawilk** [comment]
cc @PiotrSikora for correct header interpretation and @htuch as this seems security related.
I think a release note is sufficient but if anyone thinks we should send an advisory I can do that too.

File: source/common/http/conn_manager_utility.cc [hunk 1]
```diff
@@ -107,8 +110,17 @@ Network::Address::InstanceConstSharedPtr ConnectionManagerUtility::mutateRequest
         Utility::appendXff(request_headers, *connection.remoteAddress());
       }
     }
-    request_headers.insertForwardedProto().value().setReference(
-        connection.ssl() ? Headers::get().SchemeValues.Https : Headers::get().SchemeValues.Http);
+    if (trusted_forwarded_proto) {
+      // Set x-forwarded-proto unless it already exists and was forwarded on by a trusted proxy.
+      if (xff_num_trusted_hops == 0 || request_headers.ForwardedProto() == nullptr) {
```
Comment (PiotrSikora) [inline source/common/http/conn_manager_utility.cc]: Should this respect `skip_xff_append`?

Comment (alyssawilk) [reply source/common/http/conn_manager_utility.cc]: I don't _think_ so.  
skip_xff_append is to make the proxy invisible from a fowarding perspective.  We don't have a general "stop appending headers" config, we do them one at a time (there's already config knobs to avoid adding Via and I'm adding one for Server shortly).  Meanwhile if we skip adding it at all we'd break things in the router, so all this code does is decide if we set an original value or trust the downstream value, which I think is orthogonal from if we want this proxy to be a visible hop in xff at the usptream server.

Comment (PiotrSikora) [reply source/common/http/conn_manager_utility.cc]: My thinking was that `X-Forwarded-For` and `X-Forwarded-Proto` are kind of coupled together (and they are part of the same "entry" in the [Forwarded (RFC7239)](https://tools.ietf.org/html/rfc7239) header), so it would be weird seeing `X-Forwarded-Proto` alone...

But if you think that's fine, then it's fine. I just wanted to make sure you didn't miss this. The code is definitely an improvement over what we have now. Thanks!

File: test/test_common/test_runtime.h [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+// TODO(alyssawilk) move existing runtime tests over to using this.
+class TestScopedRuntime {
+public:
+  TestScopedRuntime() : api_(Api::createApiForTest()) {
+    envoy::config::bootstrap::v2::LayeredRuntime config;
+    config.add_layers()->mutable_admin_layer();
+
+    loader_ = std::make_unique<Runtime::ScopedLoaderSingleton>(Runtime::LoaderPtr{
+        new Runtime::LoaderImpl(dispatcher_, tls_, config, local_info_, init_manager_, store_,
```
Comment (mattklein123) [inline test/test_common/test_runtime.h]: nit: make_unique

File: test/test_common/test_runtime.h [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+#include "gmock/gmock.h"
+
+namespace Envoy {
+
+// TODO(alyssawilk) move existing runtime tests over to using this.
+class TestScopedRuntime {
+public:
+  TestScopedRuntime() : api_(Api::createApiForTest()) {
+    envoy::config::bootstrap::v2::LayeredRuntime config;
+    config.add_layers()->mutable_admin_layer();
```
Comment (mattklein123) [inline test/test_common/test_runtime.h]: Can you add a small comment that this is required for mergeValues() to work? This wasn't immediately clear to me.

File: source/common/http/conn_manager_utility.cc [hunk 4]
```diff
@@ -85,6 +85,9 @@ Network::Address::InstanceConstSharedPtr ConnectionManagerUtility::mutateRequest
   Network::Address::InstanceConstSharedPtr final_remote_address;
   bool single_xff_address;
   const uint32_t xff_num_trusted_hops = config.xffNumTrustedHops();
+  bool trusted_forwarded_proto =
+      Runtime::runtimeFeatureEnabled("envoy.reloadable_features.trusted_forwarded_proto");
```
Comment (mattklein123) [inline source/common/http/conn_manager_utility.cc]: Why not make this true by default? I guess do we have a documented procedure of the sequence for these type of features? false by default -> true by default -> remove old code path?

File: source/common/http/conn_manager_utility.cc [hunk 5]
```diff
@@ -85,6 +85,9 @@ Network::Address::InstanceConstSharedPtr ConnectionManagerUtility::mutateRequest
   Network::Address::InstanceConstSharedPtr final_remote_address;
   bool single_xff_address;
   const uint32_t xff_num_trusted_hops = config.xffNumTrustedHops();
+  bool trusted_forwarded_proto =
```
Comment (mattklein123) [inline source/common/http/conn_manager_utility.cc]: nit: const

File: source/common/http/conn_manager_utility.cc [hunk 6]
```diff
@@ -107,8 +110,17 @@ Network::Address::InstanceConstSharedPtr ConnectionManagerUtility::mutateRequest
         Utility::appendXff(request_headers, *connection.remoteAddress());
       }
     }
-    request_headers.insertForwardedProto().value().setReference(
-        connection.ssl() ? Headers::get().SchemeValues.Https : Headers::get().SchemeValues.Http);
+    if (trusted_forwarded_proto) {
```
Comment (mattklein123) [inline source/common/http/conn_manager_utility.cc]: nit: I think you can just fold this in below as `!trusted_forwarded_proto || ...` right?

File: source/common/http/conn_manager_utility.cc [hunk 7]
```diff
@@ -107,8 +110,17 @@ Network::Address::InstanceConstSharedPtr ConnectionManagerUtility::mutateRequest
...
+      // Set x-forwarded-proto unless it already exists and was forwarded on by a trusted proxy.
```
Comment (mattklein123) [inline source/common/http/conn_manager_utility.cc]: nit: can you rephrase this comment to read more in line with the logic that follows (i.e. or vs. and, etc.). Normally I wouldn't be picky like this but this code is so hard to understand as it is that any comments hopefully should follow pretty closely. (I'm having a bit of a hard time groking what is happening here.)

File: source/common/http/conn_manager_utility.cc (hunk 6 above)
Comment (alyssawilk) [reply source/common/http/conn_manager_utility.cc]: I can, though I lean towards not.
In practice I find it easier to verify there's no behavioral change if you have
if (guard) {
  new_behavior
} else {
  old_behavior
}
It's also easier to clean up guarded behavior by just removing an if else block.  This is simple enough I can go either way but I kind of prefer using our internal best practices for safe addition/removal of code since they evolved from a bunch of bugs where we had booleans inverted and accidentally turned things on by default and off when we flipped flags etc.

File: source/common/http/conn_manager_utility.cc (hunk 7 above)
Comment (alyssawilk) [reply source/common/http/conn_manager_utility.cc]: Sure.  Also note that I could just overwrite if num_trusted_hops is 0 and allow the logic at :139 to set proto if absent if you prefer that.  I felt it was more clear to have it in one place than fall through, but that may just be me :-)

File: source/common/http/conn_manager_utility.cc (hunk 4 above)
Comment (alyssawilk) [reply source/common/http/conn_manager_utility.cc]: Yeah, generally false -> true -> remove, but for bugfixes and low risk code I think it's fine to set true by default so done.

File: source/common/http/conn_manager_utility.cc (hunk 6 above)
Comment (mattklein123) [reply source/common/http/conn_manager_utility.cc]: Sure that's fine.

File: source/common/http/conn_manager_utility.cc (hunk 4 above)
Comment (mattklein123) [reply source/common/http/conn_manager_utility.cc]: Not related to this PR, but do your scripts track this somehow? I think this is all awesome, I just want to make sure that we don't forget about the sequence of things we need to do upon each release?

File: test/test_common/test_runtime.h (hunk 3 above)
Comment (mattklein123) [reply test/test_common/test_runtime.h]: Sorry I meant to add a comment above this line along the lines of "the existence of an admin layer is required for mergeValues() to work"

File: source/common/http/conn_manager_utility.cc (hunk 4 above)
Comment (alyssawilk) [reply source/common/http/conn_manager_utility.cc]: Yep, the scripts grep through the whole code base and add any flags which aren't already defaulted true.

File: docs/root/intro/version_history.rst [hunk 8]
```diff
@@ -25,6 +25,7 @@ Version history
 * grpc-json: added support for :ref:`ignoring unknown query parameters<envoy_api_field_config.filter.http.transcoder.v2.GrpcJsonTranscoder.ignore_unknown_query_parameters>`.
 * header to metadata: added :ref:`PROTOBUF_VALUE <envoy_api_enum_value_config.filter.http.header_to_metadata.v2.Config.ValueType.PROTOBUF_VALUE>` and :ref:`ValueEncode <envoy_api_enum_config.filter.http.header_to_metadata.v2.Config.ValueEncode>` to support protobuf Value and Base64 encoding.
 * http: added the ability to reject HTTP/1.1 requests with invalid HTTP header values, using the runtime feature `envoy.reloadable_features.strict_header_validation`.
+* http: changed Envoy to forward existing x-forwarded-proto from upstream trusted proxies. Guarded by `envoy.reloadable_features.trusted_forwarded_proto` which defaults true.
```
Comment (rgs1) [inline docs/root/intro/version_history.rst]: hmm shouldn't this be:

"from downstream trusted proxies"

?

Comment (rgs1) [reply docs/root/intro/version_history.rst]: if so: https://github.com/envoyproxy/envoy/pull/8021
